build/dart/rules.gni

```diff
@@ -168,7 +169,129 @@ template("dart_snapshot") {
   }
 }
 
-template("dart_pkg_helper") {
+# Creates an app-jit snapshot for a command-line Dart program based on a
+# training run.
+#
+# Parameters:
+#  main_dart (required):
+#    The entrypoint to the Dart application.
+#
+#  training_args (required):
+#    Arguments to pass to the Dart application for the training run.
+#
+#  vm_args (optional):
+#    Additional arguments to the Dart VM.
+#
+#  deps (optional):
+#    Any build dependencies.
+#
+#  dot_packages (required):
+#    The .packages file for the app. Defaults to the $_dart_root/.packages.
+#
+#  output (optional):
+#    Overrides the full output path.
+#
+#  snapshot_kind (optional)
+#    Either an "app-jit" snapshot (default) or a "kernel" snapshot
+template("application_snapshot") {
+  assert(defined(invoker.main_dart), "Must specify 'main_dart'")
+  assert(defined(invoker.training_args), "Must specify 'training_args'")
+  assert(defined(invoker.dot_packages), "Must specify 'dot_packages'")
+
+  main_dart = invoker.main_dart
+  training_args = invoker.training_args
+  dot_packages = rebase_path(invoker.dot_packages)
+  name = target_name
+
+  extra_deps = []
+  if (defined(invoker.deps)) {
+    extra_deps += invoker.deps
+  }
+  extra_inputs = [ main_dart ]
+  if (defined(invoker.inputs)) {
+    extra_inputs += invoker.inputs
+  }
+  output = "$root_gen_dir/$name.dart.snapshot"
+  if (defined(invoker.output)) {
+    output = invoker.output
+  }
+
+  depfile = output + ".d"
+  abs_depfile = rebase_path(depfile)
+  abs_output = rebase_path(output)
+  rel_output = rebase_path(output, root_build_dir)
+  snapshot_vm_args = [
+    "--disable-dart-dev",
+    "--deterministic",
+    "--packages=$dot_packages",
+    "--snapshot=$abs_output",
+    "--snapshot-depfile=$abs_depfile",
+    "--depfile-output-filename=$rel_output",
+  ]
+  if (defined(invoker.vm_args)) {
+    snapshot_vm_args += invoker.vm_args
+  }
+
+  snapshot_kind = "app-jit"
+  if (target_cpu != host_cpu) {
+    snapshot_kind = "kernel"
+  }
```

Comment:
nit: IMO it would be better to explicitly control this via an argument, rather than detecting it this way.

Reply:
Ahh I see you have it below - it still might be good to require people to specify it up front to avoid confusion/bad assumptions about what kind will be produced when using this template.